pennylane/capture/base_interpreter.py

```diff
@@ -0,0 +1,415 @@
+                return new_op
+
+    Now the interpreter can be used to transform functions and jaxpr:
+
+    >>> interpreter = SimplifyInterpreter()
+    >>> def f(x):
+    ...     qml.RX(x, 0)**2
+    ...     qml.adjoint(qml.Z(0))
+    ...     return qml.expval(qml.X(0) + qml.X(0))
+    >>> simplified_f = interpreter(f)
+    >>> print(qml.draw(simplified_f)(0.5)
+    0: ──RX(1.00)──Z─┤  <2.00*X>
+    >>> jaxpr = jax.make_jaxpr(f)(0.5)
+    >>> interpreter.eval(jaxpr.jaxpr, [], 0.5)
+    [expval(2.0 * X(0))]
+
+    It will also preserve higher order primitives by default:
+
+    >>> def g(x):
+    ...     @qml.for_loop(3)
+    ...     def loop(i, x):
+    ...         qml.RX(x, 0) ** i
+    ...         return x
+    ...     loop(1.0)
+    ...     return qml.expval(qml.Z(0) + 3*qml.Z(0))
+    >>> jax.make_jaxpr(interpreter(g))(0.5)
+    { lambda ; a:f32[]. let
+        _:f32[] = for_loop[
+        jaxpr_body_fn={ lambda ; b:i32[] c:f32[]. let
+            d:f32[] = convert_element_type[new_dtype=float32 weak_type=True] b
+            e:f32[] = mul c d
+            _:AbstractOperator() = RX[n_wires=1] e 0
+            in (c,) }
+        n_consts=0
+        ] 0 3 1 1.0
+        f:AbstractOperator() = PauliZ[n_wires=1] 0
+        g:AbstractOperator() = SProd[_pauli_rep=4.0 * Z(0)] 4.0 f
+        h:AbstractMeasurement(n_wires=None) = expval_obs g
+    in (h,) }
+
+
+    """
+
+    _env: dict
+    _primitive_registrations: dict["jax.core.Primitive", Callable] = {}
+    _op_math_cache: dict
+
+    def __init_subclass__(cls) -> None:
+        cls._primitive_registrations = copy.copy(cls._primitive_registrations)
+
+    def __init__(self):
+        self._env = {}
+        self._op_math_cache = {}
+
+    @classmethod
+    def register_primitive(cls, primitive: "jax.core.Primitive") -> Callable[[Callable], Callable]:
+        """Registers a custom method for handling a primitive
+
+        Args:
+            primitive (jax.core.Primitive): the primitive we want  custom behavior for
+
+        Returns:
+            Callable: a decorator for adding a function to the custom registrations map
+
+        Side Effect:
+            Calling the returned decorator with a function will place the function into the
+            primitive registrations map.
+
+        .. code-block:: python
+
+            my_primitive = jax.core.Primitive("my_primitve")
+
+            @Interpreter_Type.register(my_primitive)
+            def handle_my_primitive(self: Interpreter_Type, *invals, **params)
+                return invals[0] + invals[1] # some sort of custom handling
+
+        """
+
+        def decorator(f: Callable) -> Callable:
+            cls._primitive_registrations[primitive] = f
+            return f
+
+        return decorator
+
+    # pylint: disable=unidiomatic-typecheck
+    def read(self, var):
+        """Extract the value corresponding to a variable."""
+        if self._env is None:
+            raise ValueError("_env not yet initialized.")
```

Comment:
This will be called _a lot_, right? Maybe we should not perform this check every time?

Reply:
Trust that anyone working with these can handle some uncertainty and rely on Python erroring out on the later line? Also looks like I'm now initialization to a dict on init anyway.  Deleting this check.

Reply:
It's a leap of faith, but maybe we need that for performance. Glad it was easy to remove with the new init.